File: 200-number-of-islands/number-of-islands.cpp
class Solution {
public:
    void DFS(vector<vector<char>>& grid, int i, int j){
        // boundary checking
        if(i < 0 || i >= grid.size() || j < 0 || j >= grid[0].size())
            return;
        // return if current position is of water=='0' or is already visited=='2'
        if(grid[i][j] == '0')
            return;
        
        // mark the current as visited
        grid[i][j] = '0';

        // do DFS in all 4 directions
        DFS(grid, i+1, j);    // DOWN
        DFS(grid, i, j+1);    // RIGHT
        DFS(grid, i-1, j);    // UP
        DFS(grid, i, j-1);    // LEFT
    }

    int numIslands(vector<vector<char>>& grid) {
        int islands = 0;

        // We make each 1 as 2 in when it is visited
        for(int i = 0; i < grid.size(); i++){
            for(int j = 0; j < grid[0].size(); j++){
                // do DFS in case has not been visited and there is land
                if(grid[i][j] == '1'){
                    DFS(grid, i, j);
                    islands++;
                }
            }
        }
        return islands;
    }
};

File: 200-number-of-islands/number-of-islands.swift
class Solution {
    func dfs(_ grid: inout [[Character]], _ i: Int, _ j: Int) {
        // checks if the current cell (i, j) is within the bounds of the grid and is an unvisited land cell (char '1').
        // If any of these conditions are not met, the function immediately returns, 
        // preventing further execution of the DFS algorithm for that cell. 
        if i < 0 || i >= grid.count || j < 0 || j >= grid[0].count { return }
        if grid[i][j] != "1" { return }
        grid[i][j] = "0" // mark as visited
        
        // explore adjacent cells
        dfs(&grid, i+1, j) // down
        dfs(&grid, i-1, j) // up
        dfs(&grid, i, j+1) // right
        dfs(&grid, i, j-1) // left
    }

    func numIslands(_ grid: [[Character]]) -> Int {
        var grid = grid
        var islands = 0

        for row in 0..<grid.count {
            for col in 0..<grid[0].count {
                if grid[row][col] == "1" {
                    islands += 1
                    dfs(&grid, row, col)
                }
            }
        }
    
    return islands
    }
}